src/rust/engine/BUILD

```diff
@@ -0,0 +1,36 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+files(
+    name="rust_sources",
+    sources=[
+        "Cargo.lock",
+        "build.rs",
+        "**/Cargo.toml",
+        "**/*.rs",
+        "**/*.proto",
+        ".cargo/config",
+        "rust-toolchain",
+    ],
+)
+
+shell_sources(name="shell")
+
+shell_command(
+    name="engine-and-client",
```

Comment:
Note I'm using `shell_command` and not `adhoc_tool`.

The main differences is that `shell_command` is much more succinct albeit at the cost of not fingerprinting.
The fingerprinting is only useful for Python3.9, but in order to get that working correctly (where we'd get Python 3.9, but expose it as `python3`) is hairy, and the eventual future we want to work towards is MORE Pants-proivded-toolchains. So I decided to include the Python-3.9-via-PBS in this PR for simplicity.

Then, getting the Pants-provided-Python-3.9-toolchain working with `adhoc_tool` is doubly hard:
- We'd get the `python3` symlink in the `PATH`, but not the rest of the contents for `python3`, because `shell_command` isn't runnable
- `run_shell_command` which is runnable, doesn't have any field for "output dependencies" (which kinda makes sense) and also isn't marked as "hermetic"

So, `shell_command` with Pants-provided-Python3.9 is the clear winner.